### ISSUE TITLE
use in typescript

### ISSUE THREAD

**Issue by luwzchn** [josdejong/mathjs#990]
thanks，How can use in typescript i want。Can provide typescript version? or .d.ts ?

**Comment by firepick1** [comment]
I no longer use typescript, but I seem to remember that typescript is 100% compatible with all JS libraries, since TS transpiles into JS. If, on the other hand, you want full TS type checking for MathJS, that's a very big requirement and a lot of work.

**Comment by josdejong** [comment]
@luwzchn there are definitions files for mathjs availabe via DefinitelyTyped:

https://www.npmjs.com/package/@types/mathjs

**Comment by luwzchn** [comment]
@josdejong Great！thanks！

**Comment by josdejong** [comment]
:+1:

**Comment by ccorcos** [comment]
@josdejong have you ever used typescript? I think you'd **really** like it.

**Comment by josdejong** [comment]
Yes I've used TypeScript for one and a half year on a few large projects (Angular, React, and React Native). I'm a bit mixed: I love having intellisense on your object models and function signatures, but dealing with types also distracts from the actual logic you're writing. A few downsides I encountered is dealing with libraries having no typescript definitions or, even worse, having outdated definitions. And working with typescript+react+redux and generics gets really hairy, trying to explain to TypeScript that redux converts props into state and stuff like that. These "small" edge cases can cost a lot of time and frustration.

**Comment by ccorcos** [comment]
Outdated types are annoying, but you can always uninstall them and treat it all as unsafe Javascript...

As for Redux, I feel your pain. Redux is actually quite impure and thus difficult to type well. If you used something more along the lines of the Elm pattern, then types would be a breeze.

But in general, Typescript gets painful when you have highly polymorphic code. For me at least, the type system brings a huge degree of clarity though. I'm trying to learn more about computer algebra and reading through your code. Its very clean, but I've been spoiled lately with Typescript and I think it would bring even more clarity.

I'm curious to see how your `typed` function would look in terms of a typescript generic. :)

**Comment by josdejong** [comment]
> I'm curious to see how your typed function would look in terms of a typescript generic. :)

Have a look at this discussion: https://github.com/josdejong/typed-function/issues/10 :D

**Comment by ccorcos** [comment]
Interesting. Rather than use strings that need to be parsed to represent runtime data types, I would simply create a type generic that helps you create a data structure in parallel with the type. 

For example, [here's a simple validation library](https://github.com/ccorcos/ts-validator) I made for validating JSON at runtime. And I've built a set of composable functions for creating the validator with the exact same structure as the Typescript type. I could imagine something similar for the typed function but additionally uses string literals and structure for representing thins like union types.
